### ISSUE TITLE
No unnecessary parentheses warning in types

### ISSUE THREAD

**Issue by oli-cosmian** [rust-lang/rust#64169]
Types can be parenthesized, which is useful for e.g. `dyn (Foo + 'static)` or similar, but often these parentheses are unnecessary. Whenever a trivial type is parenthesized the compiler should suggest to remove the parentheses

```rust
pub fn foo() -> (i32) { 42 }
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=519fa5c8b61d29bc48beab3b914e4d04))

Errors:

```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.64s
```

This issue has been assigned to @sjmann via [this comment](https://github.com/rust-lang/rust/issues/64169#issuecomment-528573692).

**Comment by Centril** [comment]
I think this falls under the spirit of the `unused_parens` lint and the language team does not need to approve this as I think it is obviously right.

-----------------

Mentoring instructions:

See https://github.com/rust-lang/rust/pull/64128 for a PR that recently changed the semantics of the `unused_parens` lint. You'll want to introduce `fn check_ty` and then `match` on the [`ty.node: TyKind`](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/enum.TyKind.html) while keeping in mind to avoid suggesting the removal of parenthesis that would result in a compilation error.

**Comment by sjmann** [comment]
Would it be ok if I try to tackle this as a first time contributor? Seems like a nice starter issue.

**Comment by Centril** [comment]
@sjmann sure, go right ahead!

@rustbot assign @sjmann

**Comment by ijt** [comment]
@sjmann, are you still working on this?

**Comment by jack-t** [comment]
Hi &mdash;

I've got what I believe is a working PR over [here](https://github.com/rust-lang/rust/pull/65112). There's a kind of chicken-and-egg problem with building `core` with the lint, though. I'm not totally sure how to solve it.

Thanks!
